src/services/auth.js

```diff
@@ -1,115 +1,137 @@
-    await tokenService.saveToken(_id, tokens.refreshToken, REFRESH_TOKEN)
-
-    return tokens
-  },
-
-  sendResetPasswordEmail: async (email, language) => {
-    const user = await getUserByEmail(email)
-
-    if (!user) {
-      throw createError(404, USER_NOT_FOUND)
-    }
-
-    const { _id, firstName } = user
-
-    const resetToken = tokenService.generateResetToken({ id: _id, firstName, email })
-    await tokenService.saveToken(_id, resetToken, RESET_TOKEN)
-
-    await emailService.sendEmail(email, emailSubject.RESET_PASSWORD, language, { resetToken, email, firstName })
-  },
-
-  updatePassword: async (resetToken, password, language) => {
-    const tokenData = tokenService.validateResetToken(resetToken)
-    const tokenFromDB = await tokenService.findToken(resetToken, RESET_TOKEN)
-
-    if (!tokenData || !tokenFromDB) {
-      throw createError(400, BAD_RESET_TOKEN)
-    }
-
-    const { id: userId, firstName, email } = tokenData
-    await privateUpdateUser(userId, { password })
-
-    await tokenService.removeResetToken(userId)
-
-    await emailService.sendEmail(email, emailSubject.SUCCESSFUL_PASSWORD_RESET, language, {
-      firstName
-    })
-  }
-}
-
-module.exports = authService
+const { OAuth2Client } = require('google-auth-library')
+const tokenService = require('~/services/token')
+const emailService = require('~/services/email')
+const { getUserByEmail, createUser, privateUpdateUser, getUserById } = require('~/services/user')
+const { createError } = require('~/utils/errorsHelper')
+const {
+  EMAIL_NOT_CONFIRMED,
+  INCORRECT_CREDENTIALS,
+  BAD_RESET_TOKEN,
+  BAD_REFRESH_TOKEN,
+  USER_NOT_FOUND
+} = require('~/consts/errors')
+const emailSubject = require('~/consts/emailSubject')
+const {
+  tokenNames: { REFRESH_TOKEN, RESET_TOKEN, CONFIRM_TOKEN }
+} = require('~/consts/auth')
+const {
+  gmailCredentials: { clientId }
+} = require('~/configs/config')
+
+const authService = {
+  signup: async (role, firstName, lastName, email, password, language) => {
+    const user = await createUser(role, firstName, lastName, email, password, language)
+
+    const confirmToken = tokenService.generateConfirmToken({ id: user._id, role })
+    await tokenService.saveToken(user._id, confirmToken, CONFIRM_TOKEN)
+    await emailService.sendEmail(email, emailSubject.EMAIL_CONFIRMATION, language, { confirmToken, email, firstName })
+    return {
+      userId: user._id,
+      userEmail: user.email
+    }
+  },
+
+  googleAuth: async (idToken, role, language) => {
+    const client = new OAuth2Client(clientId)
+    const ticket = await client.verifyIdToken({
+      idToken,
+      audience: clientId
+    })
+    const { email, given_name, family_name, sub } = ticket.getPayload()
+
+    if (role) {
+      let user = await getUserByEmail(email)
+      if (!user) {
+        await createUser(role, given_name, family_name, email, sub, language, true)
+      }
+    }
+
+    return module.exports.login(email, sub, true)
+  },

```

Comment:
You can save the created user for future use

Reply:
Updated👍🏻